test/subtype.jl

```diff
@@ -0,0 +1,494 @@
+using Base.Bottom
+using Base.Test
+
+macro UnionAll(var, expr)
+    lb = :Bottom
+    ub = :Any
+    if isa(var,Expr) && var.head === :comparison
+        if length(var.args) == 3
+            v = var.args[1]
+            if var.args[2] == :(<:)
+                ub = esc(var.args[3])
+            elseif var.args[2] == :(>:)
+                lb = esc(var.args[3])
+            else
+                error("invalid bounds in UnionAll")
+            end
+        elseif length(var.args) == 5
+            v = var.args[3]
+            if var.args[2] == var.args[4] == :(<:)
+                lb = esc(var.args[1])
+                ub = esc(var.args[5])
+            else
+                error("invalid bounds in UnionAll")
+            end
+        else
+            error("invalid bounds in UnionAll")
+        end
+    elseif isa(var,Expr) && var.head === :(<:)
+        v = var.args[1]
+        ub = esc(var.args[2])
+    elseif isa(var,Expr) && var.head === :(>:)
+        v = var.args[1]
+        lb = esc(var.args[2])
+    elseif !isa(var,Symbol)
+        error("invalid variable in UnionAll")
+    else
+        v = var
+    end
+    quote
+        let $(esc(v)) = TypeVar($(Expr(:quote,v)), $lb, $ub)
+            UnionAll($(esc(v)), $(esc(expr)))
+        end
+    end
+end
+
+const issub = issubtype
+issub_strict(x::ANY,y::ANY) = issub(x,y) && !issub(y,x)
+isequal_type(x::ANY,y::ANY) = x == y
+
+# level 1: no varags, union, UnionAll
+function test_1()
+    @test issub_strict(Int, Integer)
+    @test issub_strict(Array{Int,1}, AbstractArray{Int,1})
+
+    @test isequal_type(Int, Int)
+    @test isequal_type(Integer, Integer)
+    @test isequal_type(Array{Int,1}, Array{Int,1})
+    @test isequal_type(AbstractArray{Int,1}, AbstractArray{Int,1})
+
+    @test issub_strict(Tuple{Int,Int}, Tuple{Integer,Integer})
+    @test issub_strict(Tuple{Array{Int,1}}, Tuple{AbstractArray{Int,1}})
+
+    @test isequal_type(Tuple{Integer,Integer}, Tuple{Integer,Integer})
+
+    @test !issub(Tuple{Int,Int}, Tuple{Int})
+    @test !issub(Tuple{Int}, Tuple{Integer,Integer})
+end
+
+# level 2: varargs
+function test_2()
+    @test issub_strict(Tuple{Int,Int}, Tuple{Vararg{Int}})
```

Comment:
somewhere it would be good to be very clear about whether `a <: b` implies it's true for all possible `a` and `b`. For example, `!(Tuple{Int,Int} <: Tuple{Vararg{Int,N}})` when `N=1` but it is true for any `N` bigger than 1.